source/fleetprovisioning/FleetProvisioning.cpp

```diff
@@ -674,4 +676,19 @@ bool FleetProvisioning::ExportRuntimeConfig(
     chmod(file.c_str(), S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
     FileUtils::ValidateFilePermissions(file.c_str(), Permissions::RUNTIME_CONFIG_FILE, false);
     return true;
-}
\ No newline at end of file
+}
+
+Aws::Crt::Map<Aws::Crt::String, Aws::Crt::String> FleetProvisioning::MapParameters(const Aws::Crt::String escaped_json)
+{
+    Aws::Crt::JsonObject value(escaped_json);
```

Comment:
I think there needs to be some error handling here. Based on Config.cpp, it looks like the value is read directly into the config as long as it's present. Here, if the JsonObject cannot be parsed, I believe it will pass a nullptr when you call `value.View()`. You should be able to call `WasParseSuccessful` to determine whether or not you've got something valid to work with. We should also make sure that there are unit tests that verify that the error handling works as expected.

Reply:
@matthewpaul I've improved error handling inside the MapParameters function. However, the function is private, thus I'm not sure how I can add unit tests for error handling. The function is called from within ProvisionDevice public function, but this is a complex function to unit test as it calls AWS IoT APIs.
Can you suggest how do you think it is best to add unit test support?
The latest code should be accessible using my fork here:
https://github.com/doronbl/aws-iot-device-client/blob/main/source/fleetprovisioning/FleetProvisioning.cpp

Reply:
Why does the function need to be private? It looks like with some minor modifications you could even make the method static by just returning the map instead of updating the member variable. This would open it up for testing.

Reply:
I do not think this should be exposed as an API. I've changed the code on my branch such that MapParameters is now part of the Config rather than FleetProvisioning class. It seems more correct to have it checked and mapped during configuration time rather than provisioning time.
By placing the function in Config PlainConfig::FleetProvisioning::MapParameters I run into wired situation where the code used to work in FleetProvisioning is asserting at runtime with what seems to be null allocator. I'm not able to resolve the root cause for this assertion.

I guess I'm missing some include in Config.cpp to introduce the missing allocator.

Assertion stack trace:
Fatal error condition occurred in /home/admin/git-fork/aws-iot-device-client/build/aws-iot-device-sdk-cpp-v2-src/crt/aws-crt-cpp/include/aws/crt/StlAllocator.h:49: m_allocator
Exiting Application
################################################################################
Resolved stacktrace:
################################################################################
0x000000000016b6d0: aws_backtrace_print at system_info.c:376
0x0000000000154ccc: aws_fatal_assert at assert.c:17
0x0000000000041230: Aws::Crt::StlAllocator<std::_Rb_tree_node<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > > > >::allocate(unsigned long, void const*) at StlAllocator.h:50
0x0000000000040c38: std::allocator_traits<Aws::Crt::StlAllocator<std::_Rb_tree_node<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > > > > >::allocate(Aws::Crt::StlAllocator<std::_Rb_tree_node<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > > > >&, unsigned long) at alloc_traits.h:301
0x0000000000040250: std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > >, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > > >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > >, Aws::Crt::StlAllocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > > > >::_M_get_node() at stl_tree.h:599
0x000000000003f4c0: std::_Rb_tree_node<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > > >* std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > >, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > > >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > >, Aws::Crt::StlAllocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > > > >::_M_create_node<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > >(std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> >&&) at stl_tree.h:653
0x000000000003eaac: std::pair<std::_Rb_tree_iterator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > > >, bool> std::_Rb_tree<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> >, std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > >, std::_Select1st<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > > >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > >, Aws::Crt::StlAllocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > > > >::_M_emplace_unique<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > >(std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> >&&) at stl_tree.h:2367
0x000000000003df68: std::pair<std::_Rb_tree_iterator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > > >, bool> std::map<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > >, Aws::Crt::StlAllocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > > > >::emplace<std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > >(std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> >&&) at stl_map.h:575
0x0000000000037614: Aws::Iot::DeviceClient::PlainConfig::FleetProvisioning::MapParameters(std::__cxx11::basic_string<char, std::char_traits<char>, Aws::Crt::StlAllocator<char> > const&) at Config.cpp:905 (discriminator 3)
0x0000000000036fa8: Aws::Iot::DeviceClient::PlainConfig::FleetProvisioning::LoadFromCliArgs(std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const&) at Config.cpp:849 (discriminator 6)
0x0000000000033778: Aws::Iot::DeviceClient::PlainConfig::LoadFromCliArgs(std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const&) at Config.cpp:214 (discriminator 3)
0x0000000000039a3c: Aws::Iot::DeviceClient::Config::init(std::map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > const&) at Config.cpp:1238
0x00000000000491b8: main at main.cpp:244 (discriminator 3)
0x0000ffff9172cd24: ?? ??:0
0x0000000000028874: _start at :?